### PR TITLE
chore(Data/Real/IsNonarchimedean): reuse ultrametric proofs

### PR DESCRIPTION
**Pull request by pechersky**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by pechersky** [comment]
I was unable to generalize the `multiset` proof. Ideally, we would have a lemma that stated that `‖s.prod‖ ≤ ‖s.dedup.prod‖`, but the proof evaded me.

File: Mathlib/Algebra/Order/Hom/Normed.lean [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import Mathlib.Algebra.Order.Hom.Basic
+import Mathlib.Analysis.Normed.Group.Basic
+
+/-!
+# Constructing (semi)normed groups from (semi)normed homs
+
+This file defines constructions that upgrade `(Comm)Group` to `(Semi)Normed(Comm)Group`
+using a `Group(Semi)normClass` when the codomain is the reals.
+
+See `Mathlib.Algebra.Order.Hom.Ultra` for further upgrades in to nonarchimedean normed groups.
```
Comment (mariainesdff) [inline Mathlib/Algebra/Order/Hom/Normed.lean:15]:
```suggestion
See `Mathlib.Algebra.Order.Hom.Ultra` for further upgrades in to nonarchimedean normed groups.
```
See `Mathlib.Algebra.Order.Hom.Ultra` for further upgrades to nonarchimedean normed groups.

File: Mathlib/Algebra/Order/Hom/Ultra.lean [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+The lemmas need to assume `[Dist α]` to be able to be stated, so they take an extra argument
+that shows that this distance instance is propositionally equal to the one that comes from the
+hom-based `AddGroupSeminormClass.toSeminormedAddGroup f` construction. To help at use site,
+the argument is an autoparam that resolves by definitional equality when using these constructions.
+-/
+
+variable {F α : Type*} [FunLike F α ℝ]
+
+/-- Proves that an `AddSeminormedGroup` structure constructed from an `AddGroupSeminormClass`
+that is `IsNonarchimedean` satisfies `IsUltrametricDist`. -/
```
Comment (mariainesdff) [inline Mathlib/Algebra/Order/Hom/Ultra.lean:26]:
```suggestion
/-- Proves that an `AddSeminormedGroup` structure constructed from an `AddGroupSeminormClass`
that `IsNonarchimedean` satisfies `IsUltrametricDist`. -/
```
Maybe?

File: Mathlib/Algebra/Order/Hom/Ultra.lean [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+lemma AddGroupSeminormClass.isUltrametricDist [AddGroup α] [AddGroupSeminormClass F α ℝ]
+    [inst : Dist α] {f : F} (hna : IsNonarchimedean f)
+    (hd : inst = (AddGroupSeminormClass.toSeminormedAddGroup f).toDist := by rfl):
```
Comment (mariainesdff) [inline Mathlib/Algebra/Order/Hom/Ultra.lean:29]:
```suggestion
    (hd : inst = (AddGroupSeminormClass.toSeminormedAddGroup f).toDist := by rfl) :
```

File: Mathlib/Algebra/Order/Hom/Ultra.lean [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    subst hd
+    simp_rw [dist_eq_norm, AddGroupSeminormClass.toSeminormedAddGroup_norm_eq]
+    rw [← sub_add_sub_cancel x y z]
+    exact hna _ _⟩
+
+/-- Proves that an `AddSeminormedCommGroup` structure constructed from an `AddGroupSeminormClass`
+that is `IsNonarchimedean` satisfies `IsUltrametricDist`. -/
+lemma AddGroupSeminormClass.isUltrametricDist' [AddCommGroup α] [AddGroupSeminormClass F α ℝ]
+    [inst : Dist α] {f : F} (hna : IsNonarchimedean f)
+    (hd : inst = (AddGroupSeminormClass.toSeminormedAddCommGroup f).toDist := by rfl):
```
Comment (mariainesdff) [inline Mathlib/Algebra/Order/Hom/Ultra.lean:41]:
```suggestion
    (hd : inst = (AddGroupSeminormClass.toSeminormedAddCommGroup f).toDist := by rfl) :
```

File: Mathlib/Algebra/Order/Hom/Ultra.lean (hunk 2 above)
Comment (pechersky) [reply Mathlib/Algebra/Order/Hom/Ultra.lean]: the seminorm is nonarchimedean. i'll rephrase

**Comment by mariainesdff** [comment]
LGTM, thanks!

**Pull request merged**